src/serac/physics/integrators/traction_integrator.cpp

```diff
@@ -0,0 +1,226 @@
+  dN_dxi_.SetSize(dof, dim);
+  dN_dX_.SetSize(dof, dim);
+  dxi_dX_.SetSize(dim);
+  F_.SetSize(dim);
+  Finv_.SetSize(dim);
+  nor_.SetSize(dim);
+  fnor_.SetSize(dim);
+
+  // Reshape the input state as a matrix (dof x dim)
+  input_state_matrix_.UseExternalData(input_state_vector.GetData(), dof, dim);
+
+  // Calculate an appropriate integration rule for the element order
+  int                          intorder = 2 * element_1.GetOrder() + 3;
+  const mfem::IntegrationRule& ir = mfem::IntRules.Get(parent_to_reference_face_transformation.FaceGeom, intorder);
+
+  // Initialize the residual output
+  output_residual_vector = 0.0;
+
+  for (int i = 0; i < ir.GetNPoints(); i++) {
+    // Set the current integration point
+    const mfem::IntegrationPoint& ip = ir.IntPoint(i);
+    mfem::IntegrationPoint        eip;
+    parent_to_reference_face_transformation.Loc1.Transform(ip, eip);
+    parent_to_reference_face_transformation.Face->SetIntPoint(&ip);
+
+    // Calculate the normal vector in the reference configuration
+    CalcOrtho(parent_to_reference_face_transformation.Face->Jacobian(), nor_);
+    double norm = nor_.Norml2();
+    nor_ /= norm;
+
+    if (compute_on_reference_) {
+      fnor_ = nor_;
+    } else {
+      // Calculate the deformation gradient and its inverse
+      parent_to_reference_face_transformation.Elem1->SetIntPoint(&eip);
+      CalcInverse(parent_to_reference_face_transformation.Elem1->Jacobian(), dxi_dX_);
+      element_1.CalcDShape(eip, dN_dxi_);
+      Mult(dN_dxi_, dxi_dX_, dN_dX_);
+      MultAtB(input_state_matrix_, dN_dX_, F_);
+
+      for (int d = 0; d < dim; d++) {
+        F_(d, d) += 1.0;
+      }
+
+      CalcInverse(F_, Finv_);
+
+      Finv_.MultTranspose(nor_, fnor_);
+      fnor_ *= F_.Det();
+    }
+
+    element_1.CalcShape(eip, shape_);
+    fnor_ *= pressure_.Eval(*parent_to_reference_face_transformation.Face, ip);
+
+    for (int j = 0; j < dof; j++) {
+      for (int k = 0; k < dim; k++) {
+        output_residual_vector(dof * k + j) +=
+            ip.weight * parent_to_reference_face_transformation.Face->Weight() * fnor_(k) * shape_(j);
+      }
+    }
+  }
+}
+
+void PressureIntegrator::AssembleFaceGrad(const mfem::FiniteElement& element_1, const mfem::FiniteElement& element_2,
+                                          mfem::FaceElementTransformations& parent_to_reference_face_transformation,
+                                          const mfem::Vector& input_state_vector, mfem::DenseMatrix& stiffness_matrix)
+{
+  double       diff_step = 1.0e-8;
+  mfem::Vector temp_out_1;
+  mfem::Vector temp_out_2;
+  mfem::Vector temp(input_state_vector.GetData(), input_state_vector.Size());
+
+  stiffness_matrix.SetSize(input_state_vector.Size(), input_state_vector.Size());
+
+  stiffness_matrix = 0.0;
+
+  // If computing on the deformed configuration calculate the stiffness contributions via finite difference
+  // Otherwise, the contributuion is zero
+  if (!compute_on_reference_) {
```

Comment:
Is this code called in a loop over face elements? If so, is it possible to perform this conditional outside the loop instead?

Reply:
Yes - but it would involve making a different integrator for the reference configuration application case.